m3/src/reporter_impl.h

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#pragma once
+
+#include <chrono>
+#include <condition_variable>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <set>
+#include <string>
+#include <thread>
+
+#include "thrift/protocol/TCompactProtocol.h"
+
+#include "m3/reporter.h"
+#include "m3/src/calc_transport.h"
+#include "m3/thrift/M3.h"
+#include "m3/thrift/m3_types.h"
+#include "m3/udp_transport.h"
+#include "tally/stats_reporter.h"
+
+using apache::thrift::transport::TTransport;
+
+namespace tally {
+
+namespace m3 {
+
+class Reporter::Impl : public tally::StatsReporter {
+ public:
+  Impl(const std::string &host, uint16_t port,
+       const std::map<std::string, std::string> &common_tags,
+       uint32_t max_queue_size, uint16_t max_packet_size);
+
+  ~Impl();
+
+  // Ensure the class is non-copyable.
+  Impl(const Impl &) = delete;
```

Comment:
Since we delete the copy constructor the compiler won't generate a move constructor according to [this SO post](https://stackoverflow.com/questions/23771194/why-should-i-delete-move-constructor-and-move-assignment-operator-in-a-singleton) which is the behavior we want so I think we should be okay in this regard.